File: 顶级/1001.cpp
#include<iostream>
#include "vector"
#include "algorithm"
using namespace std;
int m,n;
int MAX = 9999999;

struct Edge{
    int u,v,w,c;
};

vector<int>root;
vector<int>root_size;


bool cmp(Edge& e1,Edge& e2){
    return e1.w < e2.w;
}

int find_root(int u){
    while(root[u] != u){
        u = root[u];
    }
    return u;
}

void connect(int u,int v){
    if(root_size[u] < root_size[v]){
        root_size[v] += root_size[u];
        root[u] = v;
    }else{
        root_size[u] += root_size[v];
        root[v] = u;
    }
}


void minTree(vector<Edge> usedEdges,vector<Edge> brokenEdges){
    sort(brokenEdges.begin(),brokenEdges.end(),cmp);
    vector<int>costs = vector<int>(n+1,0);
    int max_val = 0;
    for(int i=1;i<=n;i++){
        for(int k=1;k<=n;k++){
            root[k] = k;
            root_size[k] = 1;
        }
        int part_nums = n-1;
        for(Edge e : usedEdges){
            if(e.u == i || e.v == i){
                continue;
            }
            int parent_u = find_root(e.u);
            int parent_v = find_root(e.v);
            if(parent_u != parent_v){
                connect(parent_v,parent_u);
                part_nums -= 1;
            }
        }
        for(Edge e : brokenEdges){
            if(e.u == i || e.v == i){
                continue;
            }
            int parent_u = find_root(e.u);
            int parent_v = find_root(e.v);
            if(parent_u != parent_v){
                connect(parent_v,parent_u);
                costs[i] += e.w;
                part_nums -= 1;
            }
        }
        if(part_nums > 1){
            costs[i] = MAX;
        }
        max_val = max(costs[i],max_val);
    }
    if(max_val > 0){
        bool flag = true;
        for(int i=1;i<=n;i++){
            if(costs[i] == max_val){
                if(flag){
                    flag = false;
                    cout<<i;
                }else{
                    cout<<" "<<i;
                }

            }
        }
    }else{
        cout<<0;
    }
}










int main(){
//    n =4;
//    m = 5;
//    int arr[] = {1,2,1,1,1,3,1,1,2,3,1,0,2,4,1,1,3,4,1,0};
    cin>>n>>m;
    root = vector<int>(n+1,0);
    root_size = vector<int>(n+1,1);
    vector<Edge> usedEdges;
    vector<Edge> brokenEdges;
    for(int i=0;i<m;i++){
        Edge edge =  Edge();
//        edge.u = arr[i*4];
//        edge.v = arr[i*4+1];
//        edge.w = arr[i*4+2];
//        edge.c = arr[i*4+3];
        cin>>edge.u>>edge.v>>edge.w>>edge.c;
        if(edge.c == 0){
            brokenEdges.push_back(edge);
        }else{
            usedEdges.push_back(edge);
        }
    }
    minTree(usedEdges,brokenEdges);

}

File: 顶级/1001.py
# -*- coding:utf8 -*




class Edge:

    def __init__(self,u,v,w,c):
        self.u = u
        self.v = v
        self.w = w
        self.c = c


class Union:
    def __init__(self,n):
        self.n = n
        self.tree = list(range(n+1))
        self.weights = [1]*(n+1)

    def connect(self,u,v):
        if self.weights[u] < self.weights[v]:
            self.weights[v] += self.weights[u]
            self.tree[u] = v
        else:
            self.weights[u] += self.weights[v]
            self.tree[v] = u

    def find_parent(self,u):
        while self.tree[u] != u:
            u = self.tree[u]
        return u

    def part_num(self,j):
        c = 0
        for i in range(1,self.n+1):
            if i != j and self.tree[i] == i:
                c += 1
        return c


def min_tree(n,edges):
    edges.sort(key=lambda x:x.w)
    max_val = 0
    MAX = 10000000
    cost = [0] * (n + 1)
    for i in range(1,n+1):
        union = Union(n)
        tmp = []
        for edge in edges:
            u,v,w,c = edge.u,edge.v,edge.w,edge.c
            if i in (u,v):
                continue
            if c == 1:
                parent_u = union.find_parent(u)
                parent_v = union.find_parent(v)
                if parent_v != parent_u:
                    union.connect(parent_v,parent_u)
                continue
            tmp.append(edge)
        for edge in tmp:
            u, v, w, c = edge.u, edge.v, edge.w, edge.c
            parent_u = union.find_parent(u)
            parent_v = union.find_parent(v)
            if parent_v != parent_u:
                union.connect(parent_u,parent_v)
                cost[i] += w

        if union.part_num(i) > 1:
            cost[i] = max(cost[i],MAX)

        max_val = max(cost[i],max_val)

    if max_val > 0:
        flag = True
        for i in range(1,n+1):
            if cost[i] == max_val:
                if flag:
                    print(i,end='')
                    flag = False
                else:
                    print(' '+str(i))
        return
    print(0)


def construct():
    s = """4 5
    1 2 1 1
    1 3 1 1
    2 3 1 0
    2 4 1 1
    3 4 1 0"""

    s1 = """4 5
    1 2 1 1
    1 3 1 1
    2 3 1 0
    2 4 1 1
    3 4 2 1"""

    arr = s.split('\n')
    edges = []
    n,m = [int(t) for t in arr[0].split(' ')]
    for i in range(1,m+1):
        u, v, w, c = [int(t) for t in arr[i].split(' ')]
        edges.append(Edge(u, v, w, c))
    min_tree(n,edges)


# construct()

if __name__ == '__main__':
    edges = []
    line = input()
    n,m = [int(t) for t in line.split(' ')]
    for i in range(m):
        line = input()
        u, v, w, c = [int(t) for t in line.split(' ')]
        edges.append(Edge(u, v, w, c))
    min_tree(n,edges)







